tests/Unit/HtmlProcessor/AbstractHtmlProcessorTest.php

```diff
@@ -86,10 +86,10 @@ public function reformatsHtml()
             '<head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"></head>' .
             '<body></body>' .
             '</html>';
-        $formattedHtml = "<!DOCTYPE HTML>\n" .
-            "<html>\n" .
-            '<head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"></head>' . "\n" .
-            "<body></body>\n" .
+        $formattedHtml = "<!DOCTYPE html>\n" .
+            '<html>' .
+            '<head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"></head>' .
+            '<body></body>' .
```

Comment:
I'm not satisfied there's a good reason to change the tests here.  We should permit all sorts of slightly malformed HTML just like a browser.  Because historically we have through PHP DOM and we don't want to break people's implementations...

If the new HTML5 component barfs on it, can we fall back to DOMDocument?

Reply:
This relates to https://github.com/MyIntervals/emogrifier/pull/831#discussion_r362020058

Reply:
Sorry, I was looking at the change out of context.

I'm wondering if there's any requirement that the HTML is reformatted with newlines added - I can't think of any reason.  Isn't the test more about confirming that everything apart from `<body>` (which is tested by the previous test, `renderRendersDocumentProvidedToFromDomDocument`) is functionally preserved (e.g. some superfluous whitespace _may_ be added outside of `<body>`).  @oliverklee ?

Most of this is in fact case insensitive (I think everything except `text/html; charset`).  However, allowing an alternative case representation may cause issues with some email clients that are not fully compliant (as we know, most email clients are not fully compliant).

To avoid issues, the [recommendation](https://www.w3.org/TR/html-polyglot/#doctype) for [robustness](https://www.w3.org/TR/html-polyglot/#dfn-polyglot-markup) is uppercase `DOCTYPE` and lowercase `html`.  I think we should require this, and thus may need convert the `HTML` to `html` in the `render` method.

Reply:
> I'm wondering if there's any requirement that the HTML is reformatted with newlines added 

No, there is none. That's just the way the HTML comes out of `DOMDocument`, and parsing the HTML in the tests in a formatting-agnostic way just had not been necessary in the tests yet.